client/qubesdb-cmd.c

```diff
@@ -5,79 +5,135 @@
 #include <windows.h>
 #include <getopt.h>
 #else
+#include <errno.h>
 #include <unistd.h>
 #endif
 
 #include <qubesdb-client.h>
 
-int opt_fullpath = 0;
-int opt_raw = 0;
-int opt_quiet = 0;
-int opt_watch_count = 1;
-int opt_wait = 0;
+static int opt_fullpath = 0;
+static int opt_raw = 0;
+static int opt_quiet = 0;
+static int opt_watch_count = 1;
+static int opt_wait = 0;
+static int opt_rm = 0;
 
 enum {
     DO_READ = 1,
     DO_WRITE,
     DO_RM,
     DO_MULTIREAD,
     DO_LIST,
-    DO_WATCH
+    DO_WATCH,
 } qdb_cmd;
 
-void encode_and_print_value(char *val) {
+#ifdef WIN32
+#define QUBES_SECURE_ZERO SecureZeroMemory
+#else
+#define QUBES_SECURE_ZERO explicit_bzero
+#endif
+
+static int encode_and_print_value(char *val, char buffer) {
     size_t i;
-    size_t len;
+    size_t len = strlen(val);
 
     if (opt_raw) {
-        printf("%s", val);
-        return;
+        char *old_val = val;
+        i = len + 1;
+        val[len] = '\n';
+#ifndef WIN32
+        if (buffer) {
+#endif
+            if (fwrite(val, 1, i, stdout) != i) {
+                perror("fwrite");
+                QUBES_SECURE_ZERO(old_val, len + 1);
+                return 1;
+            }
+#ifndef WIN32
+        } else do {
+            ssize_t written = write(1, val, i);
+            if (written == -1) {
+                int e = errno;
+                if (e != EAGAIN && e != EWOULDBLOCK && e != EINTR) {
+                    perror("write");
+                    QUBES_SECURE_ZERO(old_val, len + 1);
+                    return 1;
+                }
+            } else {
+                val += (size_t)written;
+                i -= (size_t)written;
+            }
+        } while (i > 0);
+#endif
+        QUBES_SECURE_ZERO(old_val, len + 1);
+        return 0;
     }
-    len = strlen(val);
     for (i = 0; i < len; i++) {
+        int status;
         if (val[i] >= 0x20 && (unsigned char)val[i] < 0x80)
-            printf("%c", val[i]);
-        else
+            status = printf("%c", val[i]);
```

Comment:
I wounder if a helper lib with a print-or-die would be useful for tools like this one.